### PR TITLE
Fix regression re: closing tabs via middle mouse click on Linux

### LINKED ISSUE #52: [Linux] Closing tab with middle mouse button pastes text in the next open tab
From atom/atom#2137

The middle mouse button should paste from clipboard on a text area. That happens. But when middle-clicking on tabs, it closes the tab and then pastes from clipboard in the next focused text area. It should not happen.

Here is a video showing the bug: https://dl.dropboxusercontent.com/u/96940816/bug.ogv

### LINKED ISSUE #461: [Linux][1.19.0-beta1] Closing tab with middle mouse pastes in next tab
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Same as #52, but this time I'm unable to reproduce this on 1.18, but able on 1.19.

### Steps to Reproduce

1. Open two files in the editor
2. Close one tab with middle mouse button
3. Observe that you have just pasted in what-ever was in your clipboard in the next file.

Make sure you have something in the clipboard.

**Expected behavior:** The tab should just be closed.

**Actual behavior:** Tab closes and content from my clipboard is pasted.

**Reproduces how often:** Always

### Versions

1.18:
```
~ ᐅ atom --version
Atom    : 1.18.0
Electron: 1.4.16
Chrome  : 53.0.2785.143
Node    : 6.5.0
```

1.19:
```
~ ᐅ atom-beta --version
Atom    : 1.19.0-beta1
Electron: 1.6.9
Chrome  : 56.0.2924.87
Node    : 7.4.0
```

### Additional Information

Related to #52

### PR DESCRIPTION
**Pull request by jasonrudolph**
Fixes atom/tabs#461
Fixes atom/tabs#52

---

On Linux, when the user performs a middle-button mouse click, Chromium fires both a `mousedown` event *and* a `paste` event. This PR teaches the `TextEditorComponent` to ignore the `paste` event.

When the user performs a middle-mouse click on a tab, we first get the `mousedown` event. In response to that event, we [close the tab and attempt to prevent Chromium's default processing for the event](https://github.com/atom/tabs/blob/ce1d92e0abb669155caa178bb71166b9f16f329a/lib/tab-bar-view.coffee#L416-L418). This prevents Chromium's default processing for the *`mousedown`* event, but then Chromium also fires a *`paste`* event, and that event pastes the clipboard's current content into the newly-focused text editor. 🤦

#14987 removed some of the middle-button mouse click handling related to pasting on Linux. This PR restores that logic and the related tests. (See `TextEditorComponent::didMouseDownOnContent(event)`.) With that change in place, since Atom already has its own logic for handling pasting, we shouldn't need to handle browser `paste` events. By ignoring the browser `paste` events on Linux, this PR fixes atom/tabs#461.

### TODO

- [x] Fix atom/tabs#461
- [x] Add test for `TextEditorComponent::didPaste(event)`
- [x] Address TODO in `TextEditorComponent::didPaste(event)` (https://github.com/atom/atom/pull/15183#discussion_r131672090)

### Test plan

- [x] Test on Ubuntu:
  - [x] Verify fix for https://github.com/atom/tabs/issues/461
  - [x] Verify the previous fix for https://github.com/atom/atom/issues/8648 is still intact
- [x] Test on macOS: Verify that the ctrl-click behavior fixed in https://github.com/atom/atom/pull/14987 is still intact

### CODE REVIEW

File: src/text-editor-component.js [hunk 1]
```diff
@@ -1549,6 +1551,11 @@ class TextEditorComponent {
     }
   }
 
+  didPaste (event) {
+    // TODO Explain the motivation for this logic
```
Comment (jasonrudolph) [inline src/text-editor-component.js]: I'll resolve this TODO before merging this PR.

**Comment by alexandernst** [comment]
Wasn't that fixed some weeks ago?

**Comment by ungb** [comment]
@alexandernst, the fix that went in 2 weeks ago was regarding middle clicking pasting twice(https://github.com/atom/atom/pull/14987) which is a different issue than Closing tab with middle mouse pastes in next tab(https://github.com/atom/tabs/issues/461)

**Comment by jasonrudolph** [comment]
@ungb: I think this is ready to 🚀. Wanna take it for a spin before I merge?

**Comment by jasonrudolph** [comment]
/fyi @nathansobo: Just a courtesy ping in case you're interested in reviewing this PR. If so, I certainly welcome it, but please don't feel obligated. :bow:

**Review by nathansobo — APPROVED** [review]

**Pull request merged**